test/filesystem.cc

```diff
@@ -135,6 +137,139 @@ TEST(Filesystem, extension) {
+
+  // There's no point in running this test if it will be single threaded.
+  if (num_threads < 2)
+    return;
+
+  // start_race_mutex protects the start_race_event and the start_race
+  // boolean. Condition variables can be awoken "spuriously", hence the
+  // start_race boolean.
+  std::mutex start_race_mutex;
+  bool start_race = false;
+  std::condition_variable start_race_event;
+
+  auto create_folder = [&](bool& ready_flag) {
+    // a simple (cheesy?) way for each thread to announce they are ready
+    ready_flag = true;
+
+    // All threads will queue up at start_race_event.wait(), only
+    // being released when the "race manager" sets start_race to true
+    // and calls start_race_event.notify_all().
+    std::unique_lock<std::mutex> start_race_lock(start_race_mutex);
+    start_race_event.wait(start_race_lock, [&] { return start_race; });
+    start_race_mutex.unlock();
+
+    bool success = filesystem::create_directories(nested_subdir);
+    EXPECT_TRUE(success);
+    EXPECT_TRUE(filesystem::exists(nested_subdir));
+  };
+
+  auto remove_folder = [&](bool& ready_flag) {
+    // a simple (cheesy?) way for each thread to announce they are ready
+    ready_flag = true;
+
+    // All threads will queue up at start_race_event.wait(), only
+    // being released when the "race manager" sets start_race to true
+    // and calls start_race_event.notify_all().
+    std::unique_lock<std::mutex> start_lock(start_race_mutex);
+    start_race_event.wait(start_lock, [&] { return start_race; });
+    start_race_mutex.unlock();
+
+    std::uintmax_t delete_count = filesystem::remove_all(base_subdir);
+    // I've found that two+ threads will claim that they've deleted the same
+    // file object. This results in double+ counting and prevents me from
+    // asserting anything related to the delete_count. All we can be sure
+    // is the following:
+    EXPECT_TRUE(!filesystem::exists(base_subdir));
+  };
+
+  const int num_iters = 50;
+  for (int j = 0; j < num_iters; j++) {
+    // concurrent folder creation
+    {
+      // Start some threads. Wait for each to declare themselves ready.
+      start_race = false;
+      std::vector<std::unique_ptr<std::thread>> threads(num_threads);
+      for (size_t i = 0; i < num_threads; i++) {
+        bool ready_flag = false;
+        threads[i].reset(new std::thread(create_folder, std::ref(ready_flag)));
+        while (!ready_flag) {
+          std::this_thread::yield();
+        }
+      }
+
+      // By this point we know all threads have been created and are ready.
+      // Announce the start of the race by first setting the "race_started"
+      // cv boolean and then calling "notify_all()".
+      start_race_mutex.lock();
+      start_race = true;
+      start_race_mutex.unlock();
+      start_race_event.notify_all();
+
+      for (size_t i = 0; i < num_threads; i++) {
+        threads[i]->join();
+      }
+
+      EXPECT_TRUE(filesystem::exists(nested_subdir));
```

Comment:
I guess this should be an `ASSERT_TRUE` (instead of `EXPECT_...`) since we don't want to proceed ahead if the dir creation did not fully succeed (else the test might fail when trying to create files below).

I guess we can look at other `EXPECT_...` statements and change them to `ASSERT_...` wherever applicable.

Reply:
Agree, thanks.